pkg/reconciler/pipelinerun/affinity_assistant.go

```diff
@@ -62,67 +59,129 @@ func (c *Reconciler) createOrUpdateAffinityAssistants(ctx context.Context, wb []
 
 		var claimTemplates []corev1.PersistentVolumeClaim
 		var claims []corev1.PersistentVolumeClaimVolumeSource
-		if w.PersistentVolumeClaim != nil {
-			claims = append(claims, *w.PersistentVolumeClaim.DeepCopy())
-		} else if w.VolumeClaimTemplate != nil {
-			claimTemplate := w.VolumeClaimTemplate.DeepCopy()
-			// PVCs Will be created by Affinity Assistant StatefulSet and will follow the naming format `<claimTemplate.Name>-<AffinityAssistantName>-0`
-			claimTemplate.Name = volumeclaim.GetPVCNameWithoutAffinityAssistant(w.VolumeClaimTemplate.Name, w, *kmeta.NewControllerRef(pr))
+
+		claimTemplate, claim := getAffinityAssistantPVCandVolumeClaimTemplate(w, pr)
+		if claimTemplate != nil {
 			claimTemplates = append(claimTemplates, *claimTemplate)
 		}
+		if claim != nil {
+			claims = append(claims, *claim)
+		}
+
+		aaName := getAffinityAssistantName(w.Name, pr.Name)
+		err := c.createOrUpdateAffinityAssistants(ctx, namespace, aaName, pr, claimTemplates, claims, unschedulableNodes)
+		errs = append(errs, err...)
+	}
+	return errorutils.NewAggregate(errs)
+}
+
+// createOrUpdateAffinityAssistantsPerPipelineRun creates an Affinity Assistant StatefulSet for each PipelineRun that
+// use a PersistentVolumeClaim volume. This is done to achieve Node Affinity for all TaskRuns that
+// in a PipelineRun. It also makes it possible for the tasks to execute parallel while sharing volume.
+// TODO: consume this function in the PipelineRun reconciler after
+// https://github.com/tektoncd/pipeline/pull/6790 is merged
+func (c *Reconciler) createOrUpdateAffinityAssistantsPerPipelineRun(ctx context.Context, wb []v1beta1.WorkspaceBinding, pr *v1beta1.PipelineRun, namespace string) error {
```

Comment:
Yeah, I think if we merge the 2 functions together, it will save some duplicate code. But in that case, we need to get https://github.com/tektoncd/pipeline/pull/6790 checked in first

Reply:
If merging https://github.com/tektoncd/pipeline/pull/6790 would result in this PR being simpler, I'd definitely favor merging that PR first. Alternatively, I think it would also be fine to introduce `type AffinityAssitantBehavior string` in this PR as well.

Reply:
Hi @lbernick , I have introduced `AffinityAssitantBehavior ` in this PR and merged the 2 functions, PTAL!